### ISSUE TITLE
Internal API to retreive a single CPU-bound task from Dispatchers.Default

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#3439]
IntelliJ IDEA platform has the following **core** API:

```
// In a blocking, non-suspendable world that is executed in ForkJoinPool

decompose() {
    // here happens decomposition of the task and .join()
}
```

Such API is composable in a way that it is possible to call `decompose` recursively from within subtasks.

It may be tempting to rewrite it with coroutines in a following manner:

```
fun decompose() = runBlocking {
    for (task in decomposition) {
        launch(Dispatchers.Default) { task() }
    }
}
```

The problem is -- this code is prone to both starvation and deadlocks -- each call to decompose from `Dispatchers.Default` occupies at least one thread in the backing executor for `joinBlocking()`, meaning that eventually, such code may deadlock.

It is not the case for the FJP-based solution, because `.join` on any FJP task not only awaits for the task to complete but also helps to execute subtasks if the joining thread is part of the target's FJP.

It doesn't seem that we can provide a full-blown FJP-like API out-of-the-box, but as the first steps we can provide an internal TBD API `getSingleDispatchersDefaultTaskOrNull`. With such API, it is possible to emulate `decompose` API and, after initial evaluation by the IDEA team, we may want to promote it to `@Delicate` API as one more way to bridge blocking calls with coroutine dispatchers.

**Comment by qwwdfsad** [comment]
I would love to know how we can implement `restoreThreadContext`.

AFAIR there is no API in Java to clean up and later recover all thread locals. 
But, given that we know the current `ThreadContextElement` (e.g. if it works for you to pass it as an argument), we can probably provide an API for isolating **these specific elements**.

**Comment by dovchinnikov** [comment]
We've recently had a starved thread situation after incorrect migration to coroutines in IJ (https://youtrack.jetbrains.com/issue/IDEA-302358), but it was using `Dispatchers.IO.limitedParallelism(xx)`. Thus I'd like to update the request to: API to retrieve a task from the current dispatcher. We are interested in `Dispatchers.Default`, `Dispatchers.IO` and `limitedParallelism` on any of them.

**Comment by dovchinnikov** [comment]
Another possible approach is to provide a single-function API with the semantics like this: 

// NB return type is intentionally Nothing
public suspend fun processEventsInCurrentThread(): Nothing {
    assert {
        ThreadLocalEventLoop.currentOrNull() != null // only allowed inside runBlocking
    }
    val thread = Thread.currentThread()
    if (thread !is CoroutineScheduler.Worker) {
        awaitCancellation()
    }
    while (true) {
        // TASK_NON_BLOCKING for CPU-bound thread or TASK_PROBABLY_BLOCKING in IO-bound thread
        val delayTime = thread.executeSingleTaskInTheCurrentMode()
        if (delayTime >= 0L) {
            delay(delayTime)
        } else {
            yield()
        }
    }
}
```

**Comment by dovchinnikov** [comment]
Attaching unfinished attempt to implement this as per request in a private conversation
[3439.patch.zip](https://github.com/Kotlin/kotlinx.coroutines/files/9961416/3439.patch.zip)

**Comment by dkhalanskyjb** [comment]
@dovchinnikov, is it true that `runSingleTaskFromCurrentSystemDispatcher` is no longer used and can be safely removed?

**Comment by knisht** [comment]
@dkhalanskyjb yes, it is true -- we solve this problem with parallelism compensation.
